drag_and_drop_v2/drag_and_drop_v2.py

```diff
@@ -280,12 +299,71 @@ def studio_submit(self, submissions, suffix=''):
         self.weight = float(submissions['weight'])
         self.item_background_color = submissions['item_background_color']
         self.item_text_color = submissions['item_text_color']
+        # Possible ValueError should be catched in _validate_submissions
+        self.max_items_per_zone = self._get_max_items_per_zone(submissions)
         self.data = submissions['data']
 
         return {
             'result': 'success',
         }
 
+    @staticmethod
+    def _get_max_items_per_zone(submissions):
+        """
+        Parses Max items per zone value coming from editor.
+
+        Does not catch ValueError intentionally to allow validation to capture this problem
+        """
+        raw_max_items_per_zone = submissions.get('max_items_per_zone', None)
+        if not raw_max_items_per_zone:
+            return None
+        else:
+            return int(raw_max_items_per_zone)
+
+    def _validate_submissions(self, submissions):
+        """
+        Validates that submissions passed are correct
+        """
+        errors = []
+
+        items = submissions['data']['items']
+
+        try:
+            max_items_per_zone = self._get_max_items_per_zone(submissions)
+        except ValueError as exc:
+            errors.append(_('Failed to parse "Max items per zone" - please check settings'))
+            logger.exception(exc)
+            return errors
+
+        if max_items_per_zone:
+            if max_items_per_zone < 0:
+                errors.append(_('"Max items per zone" should be positive, got {max_items_per_zone}').format(
+                    max_items_per_zone=max_items_per_zone
+                ))
+            else:
+                errors.extend(self._validate_item_zones(items, max_items_per_zone))
+
+        return errors
+
+    @staticmethod
+    def _validate_item_zones(items, max_items_per_zone):
+        errors = []
+        counter = Counter()
+        for item in items:
+            for zone in item.get('zones', []):
```

Comment:
@e-kolpakov This doesn't need special logic for backward compatibility because it operates on item data from `submissions` passed to `studio_submit`, right? If it were operating on existing instance data, the following scenario might cause "item zone validation" to be skipped:

A course author sets `max_items_per_zone` to a positive integer on an old DnDv2 instance with items that still use `item["zone"]` instead of `item["zones"]` and then tries to save it for the first time. Since validation happens before any of the updated data gets saved, `item.get('zones', [])` would return `[]` for each item. As a result, the item count for each zone would be `0`, and the DnDv2 instance would pass validation even if there were more than `max_items_per_zone` items associated with any of the zones.